File: app/src/main/java/com/playsho/android/utils/DimensionUtils.java
package com.playsho.android.utils;

import android.content.Context;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.graphics.Point;
import android.util.DisplayMetrics;
import android.util.Log;
import android.util.TypedValue;
import android.view.Display;
import android.view.WindowManager;

import com.playsho.android.base.ApplicationLoader;

/**
 * Utility class for handling display dimensions and density conversions.
 */
public class DimensionUtils {

    private static final String TAG = "DimensionUtils";

    // Display properties
    public static Point displaySize = new Point();
    public static boolean firstConfigurationWas;
    public static boolean usingHardwareInput;
    public static DisplayMetrics displayMetrics = new DisplayMetrics();
    public static float screenRefreshRate = 60;
    public static float density = 1;

    /**
     * Retrieves the height of the display in pixels.
     *
     * @return The height of the display.
     */
    public static int getDisplayHeightInPixel() {
        return Resources.getSystem().getDisplayMetrics().heightPixels;
    }

    /**
     * Retrieves the width of the display in pixels.
     *
     * @return The width of the display.
     */
    public static int getDisplayWidthInPixel() {
        return Resources.getSystem().getDisplayMetrics().widthPixels;
    }

    /**
     * Converts density-independent pixels (dp) to pixels (px).
     *
     * @param value The value in dp to be converted.
     * @return The converted value in pixels.
     */
    public static int dpToPx(float value) {
        if (value == 0) {
            return 0;
        }
        return (int) Math.ceil(density * value);
    }

    /**
     * Converts pixels (px) to density-independent pixels (dp).
     *
     * @param px The value in pixels to be converted.
     * @return The converted value in dp.
     */
    public static float pxToDp(int px) {
        DisplayMetrics metrics = ApplicationLoader.getAppContext().getResources().getDisplayMetrics();
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_PX, px, metrics);
    }

    /**
     * Checks and updates the display size based on the given context and configuration.
     *
     * @param context           The context.
     * @param newConfiguration  The new configuration (can be null).
     */
    public static void checkDisplaySize(Context context, Configuration newConfiguration) {
        try {
            DimensionUtils.density = context.getResources().getDisplayMetrics().density;
            firstConfigurationWas = true;
            Configuration configuration = newConfiguration;
            if (configuration == null) {
                configuration = context.getResources().getConfiguration();
            }
            usingHardwareInput = configuration.keyboard != Configuration.KEYBOARD_NOKEYS && configuration.hardKeyboardHidden == Configuration.HARDKEYBOARDHIDDEN_NO;
            WindowManager manager = (WindowManager) context.getSystemService(Context.WINDOW_SERVICE);
            if (manager != null) {
                Display display = manager.getDefaultDisplay();
                if (display != null) {
                    display.getMetrics(displayMetrics);
                    display.getSize(displaySize);
                    screenRefreshRate = display.getRefreshRate();
                }
            }
            if (configuration.screenWidthDp != Configuration.SCREEN_WIDTH_DP_UNDEFINED) {
                int newSize = (int) Math.ceil(configuration.screenWidthDp * DimensionUtils.density);
                if (Math.abs(displaySize.x - newSize) > 3) {
                    displaySize.x = newSize;
                }
            }
            if (configuration.screenHeightDp != Configuration.SCREEN_HEIGHT_DP_UNDEFINED) {
                int newSize = (int) Math.ceil(configuration.screenHeightDp * DimensionUtils.density);
                if (Math.abs(displaySize.y - newSize) > 3) {
                    displaySize.y = newSize;
                }
            }

        } catch (Exception e) {
            Log.e(TAG, "checkDisplaySize: " );
        }
    }
}

File: app/src/main/java/com/playsho/android/utils/DimensionUtils.kt
package com.playsho.android.utils
import android.content.Context
import android.content.res.Configuration
import android.content.res.Resources
import android.graphics.Point
import android.util.DisplayMetrics
import android.util.Log
import android.util.TypedValue
import android.view.WindowManager
import com.playsho.android.base.ApplicationLoader
import kotlin.math.abs
import kotlin.math.ceil

/**
 * Utility class for handling display dimensions and density conversions.
 */
object DimensionUtils {

    private const val TAG = "DimensionUtils"

    // Display properties
    var displaySize = Point()
    var firstConfigurationWas: Boolean = false
    var usingHardwareInput: Boolean = false
    var displayMetrics = DisplayMetrics()
    var screenRefreshRate: Float = 60f
    var density: Float = 1f

    /**
     * Retrieves the height of the display in pixels.
     *
     * @return The height of the display.
     */
    fun getDisplayHeightInPixel(): Int {
        return Resources.getSystem().displayMetrics.heightPixels
    }

    /**
     * Retrieves the width of the display in pixels.
     *
     * @return The width of the display.
     */
    fun getDisplayWidthInPixel(): Int {
        return Resources.getSystem().displayMetrics.widthPixels
    }

    /**
     * Converts density-independent pixels (dp) to pixels (px).
     *
     * @param value The value in dp to be converted.
     * @return The converted value in pixels.
     */
    fun dpToPx(value: Float): Int {
        if (value == 0f) {
            return 0
        }
        return ceil((density * value).toDouble()).toInt()
    }

    /**
     * Converts pixels (px) to density-independent pixels (dp).
     *
     * @param px The value in pixels to be converted.
     * @return The converted value in dp.
     */
    fun pxToDp(px: Int): Float {
        val metrics = ApplicationLoader.getAppContext().resources.displayMetrics
        return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_PX, px.toFloat(), metrics)
    }

    /**
     * Checks and updates the display size based on the given context and configuration.
     *
     * @param context The context.
     * @param newConfiguration The new configuration (can be null).
     */
    fun checkDisplaySize(context: Context, newConfiguration: Configuration?) {
        try {
            density = context.resources.displayMetrics.density
            firstConfigurationWas = true
            val configuration = newConfiguration ?: context.resources.configuration
            usingHardwareInput = configuration.keyboard != Configuration.KEYBOARD_NOKEYS && configuration.hardKeyboardHidden == Configuration.HARDKEYBOARDHIDDEN_NO
            val manager = context.getSystemService(Context.WINDOW_SERVICE) as WindowManager?
            manager?.defaultDisplay?.let { display ->
                display.getMetrics(displayMetrics)
                display.getSize(displaySize)
                screenRefreshRate = display.refreshRate
            }
            if (configuration.screenWidthDp != Configuration.SCREEN_WIDTH_DP_UNDEFINED) {
                val newSize = (configuration.screenWidthDp * density + 0.5f).toInt()
                if (abs(displaySize.x - newSize) > 3) {
                    displaySize.x = newSize
                }
            }
            if (configuration.screenHeightDp != Configuration.SCREEN_HEIGHT_DP_UNDEFINED) {
                val newSize = (configuration.screenHeightDp * density + 0.5f).toInt()
                if (abs(displaySize.y - newSize) > 3) {
                    displaySize.y = newSize
                }
            }
        } catch (e: Exception) {
            Log.e(TAG, "checkDisplaySize: ", e)
        }
    }
}